pageserver/src/billing_metrics.rs

```diff
@@ -0,0 +1,278 @@
+                    metric: BillingMetricKind::WrittenSize.to_string(),
+                },
+                timeline_written_size,
+            ));
+
+            let timeline_size = timeline.get_physical_size();
+            tenant_physical_size += timeline_size;
+
+            debug!(
+                "per-timeline current metrics for tenant: {}: timeline {} physical_size={} last_record_lsn {} (as bytes)",
+                tenant_id, timeline.timeline_id, timeline_size, timeline_written_size)
+        }
+
+        let tenant_remote_size = tenant.get_remote_size().await?;
+        debug!(
+            "collected current metrics for tenant: {}: state={:?} tenant_physical_size={} remote_size={}",
+            tenant_id, tenant_state, tenant_physical_size, tenant_remote_size
+        );
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::PhysicalSize.to_string(),
+            },
+            tenant_physical_size,
+        ));
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::PhysicalSize.to_string(),
+            },
+            tenant_remote_size,
+        ));
+
+        // TODO add SyntheticStorageSize metric
+    }
+
+    // Filter metrics
+    let mut cached_metrics_guard = cached_metrics.lock().await;
+    let mut filtered_metrics = Vec::new();
+
+    for (curr_key, curr_val) in current_metrics.iter() {
+        if let Some(val) = cached_metrics_guard.insert(curr_key.clone(), *curr_val) {
+            if val != *curr_val {
+                // metric was updated, send it
+                filtered_metrics.push({
+                    BillingMetric::new_absolute(
+                        curr_key.metric.clone(),
+                        curr_key.tenant_id,
+                        curr_key.timeline_id,
+                        *curr_val,
+                    )
+                });
+            }
+        } else {
+            // cache the metric
+            cached_metrics_guard.insert(curr_key.clone(), *curr_val);
+            filtered_metrics.push({
+                BillingMetric::new_absolute(
+                    curr_key.metric.clone(),
+                    curr_key.tenant_id,
+                    curr_key.timeline_id,
+                    *curr_val,
+                )
+            });
+        }
+    }
+
+    if filtered_metrics.is_empty() {
+        trace!("no new metrics to send");
+        return Ok(());
+    }
+
+    // Send metrics to billing service.
+    // split into chunks of 1000 metrics to avoid exceeding the max request size
+    const CHUNK_SIZE: usize = 1000;
+    let chunks = filtered_metrics.chunks(CHUNK_SIZE);
+    for chunk in chunks {
+        let chunk_json: serde_json::Value = serde_json::json!({
+            "events": serde_json::to_string(&chunk)?});
```

Comment:
In python test it is expected to parse:

```json
{ "events": [...] }
```

But I think this will produce

```json
{ "events": "[...]" }
```

With this kind of type, I think one could handle outputting an actual json array:

```rust
#[derive(serde::Serialize)]
struct EventChunk<'a> {
  events: &'a [BillingMetric]
}


for chunk in chunks {
  // then use as serde::Serialize as is currently
  let chunk_json = serde_json::value::to_raw_value(&EventChunk {
    events: chunk
  }).expect("BillingMetric should not fail serialization");
  // snip
}
```

I think there could be a `Bytes` based buffer-reusing way to serialize these, but if the payload is currently wrong this should help forward.

Reply:
I guess this got incorporated in [d9e388f](https://github.com/neondatabase/neon/pull/3012/commits/d9e388f9eff772c24d9a06c48eff95ea4e380ffc)